### PR TITLE
feat(websocket): use proxy web socket on chromium

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: src/server/chromium.ts [hunk 1]
```diff
@@ -183,6 +176,130 @@ export class Chromium implements BrowserType<CRBrowser> {
...
+      // If message has no sessionId, look it up.
+      const session = socketToBrowserSession.get(socket)!;
+      if (session.sessionId) {
+        // We have it, use it.
+        parsedMessage.sessionId = session.sessionId;
+        transport.send(parsedMessage);
+        return;
+      }
+      // Pending session id, queue the message.
+      session!.queue!.push(parsedMessage);
```
Comment (dgozman) [inline src/server/chromium.ts]: nit: no need for `session!`

**Review by dgozman — APPROVED** [review]

File: src/firefox/ffConnection.ts [hunk 2]
```diff
@@ -76,13 +76,13 @@ export class FFConnection extends platform.EventEmitter {
...
+  _rawSend(message: ProtocolRequest) {
     if (this._debugProtocol.enabled)
       this._debugProtocol('SEND ► ' + rewriteInjectedScriptEvaluationLog(message));
     this._transport.send(message);
   }
 
-  async _onMessage(message: ProtocolMessage) {
+  async _onMessage(message: ProtocolResponse) {
     if (this._debugProtocol.enabled)
       this._debugProtocol('◀ RECV ' + message);
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply src/firefox/ffConnection.ts]: It will, fixing.

File: src/server/chromium.ts [hunk 3]
```diff
@@ -114,8 +114,8 @@ export class Chromium implements BrowserType<CRBrowser> {
         // Note that it's fine to reuse the pipe transport, since
         // our connection ignores kBrowserCloseMessageId.
         const t = transport || await platform.connectToWebsocket(browserWSEndpoint!, async transport => transport);
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply src/server/chromium.ts]: Done.

File: src/server/chromium.ts [hunk 4]
```diff
@@ -125,18 +125,11 @@ export class Chromium implements BrowserType<CRBrowser> {
...
+    browserServer = new BrowserServer(launchedProcess, gracefullyClose, launchType === 'server' ? wrapTransportWithWebSocket(transport, port || 0) : null);
+    if (launchType === 'server')
       return { browserServer };
-    } else {
-      // For local launch scenario close will terminate the browser process.
-      transport = new PipeTransport(launchedProcess.stdio[3] as NodeJS.WritableStream, launchedProcess.stdio[4] as NodeJS.ReadableStream, () => browserServer!.close());
-      browserServer = new BrowserServer(launchedProcess, gracefullyClose, null);
-      return { browserServer, transport };
-    }
+    return { browserServer, transport };
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply src/server/chromium.ts]: Done.

File: src/server/webkit.ts [hunk 5]
```diff
@@ -123,8 +121,10 @@ export class WebKit implements BrowserType<WKBrowser> {
...
 
     // For local launch scenario close will terminate the browser process.
+    let transport: ConnectionTransport | undefined = undefined;
+    let browserServer: BrowserServer | undefined = undefined;
     transport = new PipeTransport(launchedProcess.stdio[3] as NodeJS.WritableStream, launchedProcess.stdio[4] as NodeJS.ReadableStream, () => browserServer!.close());
-    browserServer = new BrowserServer(launchedProcess, gracefullyClose, launchType === 'server' ? await wrapTransportWithWebSocket(transport, port || 0) : null);
+    browserServer = new BrowserServer(launchedProcess, gracefullyClose, launchType === 'server' ? wrapTransportWithWebSocket(transport, port || 0) : null);
     if (launchType === 'server')
       return { browserServer };
     return { browserServer, transport };
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply src/server/webkit.ts]: Done

File: src/server/chromium.ts [hunk 6]
```diff
@@ -183,6 +176,130 @@ export class Chromium implements BrowserType<CRBrowser> {
...
+      }
+      // Pending session id, queue the message.
+      session!.queue!.push(parsedMessage);
+    });
+
+    socket.on('close', (socket as any).__closeListener = () => {
+      const session = socketToBrowserSession.get(socket);
+      if (!session)
+        return;
+      sessionToSocket.delete(session.sessionId!);
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply src/server/chromium.ts]: Done

File: src/server/chromium.ts [hunk 7]
```diff
@@ -183,6 +176,130 @@ export class Chromium implements BrowserType<CRBrowser> {
...
+  const browserSessions = new Set<string>();
+  let lastSequenceNumber = 1;
+
+  transport.onmessage = message => {
+    if (typeof message.id === 'number' && awaitingBrowserTarget.has(message.id)) {
+      const freshSocket = awaitingBrowserTarget.get(message.id)!;
+      awaitingBrowserTarget.delete(message.id);
+
+      const sessionId = message.result.sessionId;
+      if (freshSocket.readyState !== ws.CLOSING) {
```
[the remark this answers is not in the document]
Comment (pavelfeldman) [reply src/server/chromium.ts]: Done

File: src/server/chromium.ts (hunk 1 above)
Comment (pavelfeldman) [reply src/server/chromium.ts]: Done

**Pull request merged**